Nodock Plugin Reworked

LazDisco opened the pull request:
Under the new format, /nodock works by loading in different lists of bases, depending on the ID that initialised the nodock. Like the old plugin, police IDs can still restrict all civilian/military bases with nodock, as not to rob them of their main perk, but now each ID can block the bases of it's own IFF. So Outcasts can now nodock people to Outcast bases and Liberty Navy can nodock people to Liberty navy bases - and so on.

I've already filled out the entire of the config file and it's very easy to modify. In addition to the nodock function, it is possible to have certain IDs out right restricted from certain bases or solars. While I chose no to add any, it is an option. Helpful for events and RP developments. (Perhaps a non-hostile FR5 could be done with this)

Certain ships can be rendered exempt from all nodocks as shown inside of the config.

dsyalex commented on Binaries/bin-conf/flhook_plugins/laz_NoDock.cfg:
```diff
@@ -0,0 +1,973 @@
+nodock = LI15_01
+nodock = LI15_01b
+;Guard Stations
+Li05_02
```
what's going on here?

dsyalex commented on Binaries/bin-conf/flhook_plugins/laz_NoDock.cfg:
```diff
@@ -0,0 +1,973 @@
+nodock = RH11_01
+nodock = IW13_08
+;Guard Stations
+HI18_04
```
?

LazDisco replied:
I was initially going to outright restrict guard stations so I left them out, but then I realised that might be a bad idea for RP, but forgot to go and add the nodock comments.

dsyalex replied:
Those are keys but okay

dsyalex commented on Binaries/bin-conf/flhook_plugins/laz_NoDock.cfg:
```diff
@@ -0,0 +1,973 @@
+nodock = HI03_04
+nodock = GA06_02
+;Guard Stations
+HI10_01
```
?

dsyalex commented on Plugins/Public/NoDock/Main.cpp:
```diff
@@ -0,0 +1,574 @@
+
+	char szCurDir[MAX_PATH];
+	GetCurrentDirectory(sizeof(szCurDir), szCurDir);
+	string NoDockCfg = string(szCurDir) + R"(\flhook_plugins\laz_NoDock.cfg)"; // Our Config File
```
waho what is going on here?

LazDisco replied:
The entire thing is in a new project since you mentioned you desired projects to be separated rather than merged. This seemed like a good opportunity.

dsyalex replied:
Yeah my point was the raw string thing. I hadn't seen it before, but it's okay.

dsyalex commented on Plugins/Public/NoDock/Main.cpp:
```diff
@@ -0,0 +1,574 @@
+	bool correctID = false;
+	uint noDocker; // The ID that preformed the nodock
+
+	for (auto& mapShip : mapShips)
```
this block looks like it could be turned into a direct lookup

dsyalex commented on Plugins/Public/NoDock/Main.cpp:
```diff
@@ -0,0 +1,574 @@
+
+		wstring wscTargetCharname = (const wchar_t*)Players.GetActiveCharacterName(iClientIDTarget);
+
+		for (auto& i : mapActiveNoDocks)
```
as above

dsyalex commented on Plugins/Public/NoDock/Main.cpp:
```diff
@@ -0,0 +1,574 @@
+		{
+			for (auto& iter : mapIDRestricted)
+			{
+				if(iter.first == i->iArchID)
```
looks like another direct lookup case

LazDisco commented:
The config was generated via a bit of code I wrote. It when through the game and extracted all to of the bases and their reps. I then used a regex expression to get only the bases with the reps of the faction I was currently adding to the config. 

For this reason, I haven't tested all the bases directly myself, but I am fairly sure the the code worked based on the few that I did test. I will use the script again to generate base names with it so we can comment all of the names.

dsyalex commented:
I think I'll need a copy of that code to review and run myself to generate the actual config

zibbizor commented:
If you are using external tools to generate config files it may be a good time for us to create a new folder (or repo?) to house these projects. The plugins are a bit useless if they're a config nightmare like this is.

It may be time to consider using a naming convention as well.

LazDisco commented:
https://github.com/DiscoveryGC/Tools/blob/master/Config%20Creator/NoDock/Program.cs

I wrote a program to generate it, and moved it to a shared repo. This tool isn't 100% accurate due to how Discovery does stuff, but it gets you most of the way there.

dsyalex commented:
So are you saying that it won't actually produce this config?

LazDisco commented:
> So are you saying that it won't actually produce this config?

I have just updated the generation utility to be significantly more accurate and do more of the leg work in actually generating it. There are a few factions still being left out for reasons I am trying to figure out; HF, AI, Coalition, are good examples.

The config I have just updated is almost exactly like the one the tool generates and is server usable as it stands. I'll work on covering a few of the left out factions, but for now it's a minor amount of manual editing.

LazDisco commented:
I have deleted my FLHook fork and am now unable to make changes. Closing request.
